## 中间件
中间件middleware，是一个采用责任链模式夹杂在我们的服务器方法中间的方法。

作用：在访问指定接口之前或之后的一个自定义方法，就是中间件

特点：
- 既然叫中间件，它是可以夹在方法中间的
- 责任链：这是个设计模式，调用到自己方法时，如符合处理条件，就自行处理并终止操作，否则就传递给下一个处理程序。
- 方法（function）：中间件依然是个方法
- 有序：中间件代码是有顺序的，放的位置不一样，调用时机就不同

语法：
```js
app.use((req, res, next)=>{
    // 中间件逻辑
    next()
})
// 或
router.use((req, res, next)=>{
    // 中间件逻辑
    next(someParams)
})
```
方法签名最后一个参数一定是 ```next```，且调用next就执行下一个方法，否则就需要在此处res.end否则程序会挂起

```next```方法调用时，可以接受一个Error对象作为参数，用作抛出异常，任何异常将在下一个中间件的第一个参数中获得，接下来的异常处理就会用到

## 中间件的作用
1. 充当请求的拦截器
2. 异常处理

## 异常处理
异常处理在实际项目中非常常见，因为通常程序除了异常，服务器都需要一个统一的返回数据让页面显示，为了更友好，更好看，且数据格式需要在异常处理的时候格式化成需要的数据格式。

```js
// 自定义返回结构
let result = {
    code: 0000 // 业务码
    msg: '业务成功',
    result: {}
}
```

异常处理的方式就是使用中间件，中间件的位置需要在所有请求和中间件之后书写。如：
```js
app.use((req, res, next)=>{
    console.log('this is middleware')
    next()
})

app.get('/', (req, res, next) => {
    // 接口内也能接收一个 next
    // next 函数用于接口抛出异常给后续中间件
    // console.log('this is /')
    // res.end('hello home!')
    next(new Error('my error'))
})

// 定义异常处理器
app.use((err, req, res, next)=>{
    console.log(err)
    res.status(500)
    res.end(err.stack)
})
```

> 异常处理器的方法签名中，第一个参数，始终是error，如果程序没有抛错的话，那么错误处理程序将不会被调用

#### 异常分类与处理
异常通常分为两类：系统异常，业务异常
- 系统异常：通常由于开发人员的bug导致，这种异常不应该让用户看见，所以处理的时候通常会返回："服务器忙，请稍后再试"等友好的提示语句。
- 业务异常：设计某一个具体功能，由于用户操作导致的异常。如：登录的时候密码不对，邮箱格式不正确，用户的非法输入和不和逻辑的操作都是业务异常。此类异常通常需要告知用户，用户哪里做错了来作为提示。

这两种异常怎么处理呢？通常我们不会去修改http的status code，也就是http协议的状态码。而是服务器规定一组错误码，前端解析或服务器直接返回错误提示，如下：

先判断异常是系统异常还是业务异常，这里我们需要先定义一个我们自己的异常类，如下：
```js
class MyError extends Error {

}

module.exports = {MyError}
```
然后在业务逻辑出错时，我们应该主动抛出自定义异常，如下
```js
const {MyError} = require('./myError')

app.post('/user/login', (req, res)=>{
    // 这里如果说，我们判断到用户密码不对，那么需要抛出一个业务异常
    throw new MyError('登录密码不正确')
})
```
在异常抛出后，我们需要一个异常处理逻辑来处理异常：
```js
app.use((err, req, res, next)=>{
    if (err instanceof MyError) {
        // 在这里做业务异常处理的事情
        res.end(err.message)
    } else {
        // 在这里做系统异常处理的事情
        res.end('this is not myError')
    }
})
```

#### 总结

异常处理中间件可以捕获以下异常：

1. 在接口回调函数中抛出的异常，如：`throw new Error('my error')`
2. 在接口回调函数中使用 next 函数抛出的异常，如: `(req, res, next)=>{ next(new Error('my error')) }`

异常处理中间件无法捕获的内容如下：

1. promise 中 reject 的异常
2. 若接口回调是一个异步函数，也无法捕获

处理上述无法捕获异常的方法如下:

1. 安装 `express-async-handler`
2. 给接口回调函数包装一层 `asyncHandler`

   ```js
   const asyncHandler = require('express-async-handler');
   app.get('/asyncHandler', asyncHandler(async (req, res) => {
       await Promise.reject('my error');
       res.json({code: 200, msg: 'promise 调用成功', result: null});
   }));
   ```
